pkg/hydrator/asset_vulnerability_hydrator.go

```diff
@@ -0,0 +1,154 @@
+package hydrator
+
+import (
+	"fmt"
+	"strings"
+	"sync"
+
+	"github.com/asecurityteam/nexpose-vuln-hydrator/pkg/domain"
+)
+
+// AssetVulnerabilityHydrator implements the domain.Hydrator interface for hydrating
+// an asset's vulnerabilities
+type AssetVulnerabilityHydrator struct {
+	Client domain.NexposeClient
+}
+
+// HydrateVulnerabilities takes a domain.Asset and fetches its vulnerabilities to be further
+// hydrated concurrently
+func (a *AssetVulnerabilityHydrator) HydrateVulnerabilities(asset domain.Asset) (domain.AssetVulnerabilityDetails, error) {
+	// For an asset - we first get the list of vulnerabilities for a given asset ID
+	// For each vulnerability, we can fetch these concurrently, as seen on lines 35-41
+	// For each vulnerability, we can fetch the details and solutions concurrently, as seen on lines 75-102
+	// The solutions response gives a list of solution IDs, we can again fetch the solution details concurrently as seeon on lines 134-153
+
+	// 1. fetch vulnerabilities to be hydrated - this contains contains the vulnerability ID (for further hydration)
+	// as well as the results (port and protocol information of finding)
+	vulnerabilities, err := a.Client.GetAssetVulnerabilities(asset.ID)
+	if err != nil {
+		return domain.AssetVulnerabilityDetails{}, err
+	}
+
+	vulnDetailsChan := make(chan domain.VulnerabilityDetails, len(vulnerabilities))
+	errChan := make(chan error)
+
+	var wg sync.WaitGroup
+	wg.Add(len(vulnerabilities))
+	for _, vuln := range vulnerabilities {
```

Comment:
what's the expected behavior if an asset has 100 vulnerabilities and 1 of them results in an error? in terms of vulnbutler, sending a report would make it appear as though the vulnerability was closed, but this is at risk of leaving 99 tickets un-opened for some amount of time - cc @faldridge

Reply:
Ugh, yeah, I discussed this in a pending comment, sorry. I'll add it here now: 

Reporting false remediations is end-user-facing, so I think the confusing UX here should be avoided. Stopping here and alerting the dev team avoids that. The delay of finding and resolving non-retryable errors is small amount of time in the vuln scanning context.